src/web/reducers/flows.js

```diff
@@ -160,6 +103,60 @@ function updateCurrentHash(state) {
   return { ...state, currentHashes: computeFlowsHash(state) }
 }
 
+function createSnapshot(state) {
+  const snapshot = {
+    activeFlow: state.currentFlow,
+    activeFlowNode: state.currentFlowNode,
+    flowsByName: Object.assign({}, state.flowsByName)
+  }
+
+  const lastSnapshot = _.head(state.snapshots)
+
+  let snapshots = _.take(state.snapshots, SNAPSHOT_SIZE)
+
+  if (
+    state.currentSnapshotIndex === 0 &&
+    state.snapshots.length > 1 &&
+    lastSnapshot &&
+    snapshot.activeFlow === lastSnapshot.activeFlow &&
+    (!!snapshot.activeFlowNode && snapshot.activeFlowNode === lastSnapshot.activeFlowNode)
+  ) {
+    snapshots = _.drop(snapshots, 1) // We merge the current and last snapshots
+  }
+
+  return {
+    ...state,
+    snapshots: [snapshot, ...snapshots],
+    currentSnapshotIndex: 0
+  }
+}
+
+function applySnapshot(state, snapshot) {
+  return {
+    ...state,
+    currentFlow: snapshot.activeFlow,
+    currentFlowNode: snapshot.activeFlowNode,
+    flowsByName: snapshot.flowsByName
+  }
+}
+
+function copyName(siblingNames, nameToCopy) {
+  let copies = siblingNames.filter(name => name.startsWith(`${nameToCopy}-copy`))
+
+  if (!copies.length) {
+    return `${nameToCopy}-copy`
+  }
+
+  let i = 1
+  while (true) {
```

Comment:
can instead be

```
let i = 1
while (copies.find(name => name === `${nameToCopy}-copy-${i}`)) {
  i++
}
return `${nameToCopy}-copy-${i}`
```

Reply:
This one was only moved to the bottom due to new eslint-rules, so I think we could refactor it out of scope of current task.

Reply:
Agreed